quinn-proto/src/connection/mod.rs

```diff
@@ -532,19 +613,46 @@ where
                 );
             }
 
-            if !coalesce || buf.capacity() - buf.len() < MIN_PACKET_SPACE {
+            if concat_datagram {
+                // Write another packet in a fresh datagram in the same packet space.
+                // This requires the packet to have been padded.
+                assert_eq!(buf.len(), buf_capacity, "Packet must be padded");
+
+                // Resize the buffer so that we space for another datagrams
+                buf_capacity += self.mtu as usize;
+                if buf.capacity() < buf_capacity {
+                    buf.reserve(buf_capacity - buf.capacity());
+                }
+
+                num_datagrams += 1;
+
+                // Note that trying to write another datagram might yield an
+                // datagram which purely contains a packet header if no payloa
+                // is available anymore. That could be improved by either
+                // filtering empty packets in `finish_packet` or by making
+                // sure `has_more` is more precise and doesn't yield false
+                // positives.
+
+                continue;
+            }
+
+            if !coalesce || buf_capacity - buf.len() < MIN_PACKET_SPACE {
                 break;
             }
+
+            space_idx += 1;
         }
 
         if buf.is_empty() {
             return None;
         }
 
+        num_datagrams += 1;
+
         trace!("sending {} byte datagram", buf.len());
```

Comment:
Wording of this should be tweaked.

Reply:
what do you recommend?

Reply:
Is this about the `trace!`?

Reply:
Right. Moving it up into the loop would be fine, or adjusting it to something lke "sent {} {} byte datagrams".